### PR TITLE
Add macros

### PR DESCRIPTION
**Pull request by blyxxyz**
This adds vim-style `q`/`@` macros, as in #1981. It uses the `MarkKey` system to record a single keypress but it also supports giving a register name as a command argument to `:record-macro` and `:run-macro`.

(See #1981)

### CODE REVIEW

**Comment by The-Compiler** [comment]
Thank you! I'm a bit worried about all those globals and the "pollution" of `utilcmds.py` with macro-related stuff though.

I think it'd be better to move everything macro-related to a separate module, say `qutebrowser.keyinput.macros` or so. Then we could have an object like `MacroRecorder` there, which is initialized and inserted into `objreg` in a `macros.init()` method (called from `app.py`). That can then hold all related state and the commands - how does that sound?

**Comment by blyxxyz** [comment]
Sounds good. I hope I did it right.

**Comment by The-Compiler** [comment]
Thanks! Looks much cleaner from a very quick look, I'll do another review somewhen next week (quite busy until then I'm afraid).

**Comment by HolySmoke86** [comment]
Running a macro with the `@` + key binding on an unregistered key triggers a `CommandError` that is not caught. The same works fine for named macros.
Adding the following tests to `misc.feature` exposes this behaviour:

```
    Scenario: Running an invalid macro
        Given I open data/scroll/simple.html
        And I run :tab-only
        When I run :run-macro
        And I press the key "b"
        Then the error "No macro recorded in 'b'!" should be shown
        And qutebrowser should not crash

    Scenario: Running an invalid named macro
        Given I open data/scroll/simple.html
        And I run :tab-only
        When I run :run-macro bar
        Then the error "No macro recorded in 'bar'!" should be shown
        And qutebrowser should not crash
```

Other than that, the intended functionality of recording and replaying macros works like a charm.

**Comment by blyxxyz** [comment]
That happened because `run_macro` gets called by `run_macro_command` within `run_safely` if you pass the macro name as a command, but otherwise it gets called by `RegisterKeyParser.handle` which doesn't have `run_safely` to catch its errors. Thank you.

File: qutebrowser/commands/runners.py [hunk 1]
```diff
@@ -259,19 +259,33 @@ def run(self, text, count=None):
...
             result.cmd.run(self._win_id, args, count=count)
 
-            if result.cmdline[0] != 'repeat-command':
-                last_command[cur_mode] = (text, count)
+            if result.cmdline[0] == 'repeat-command':
+                record_last_command = False
+
+            if result.cmdline[0] in ['record-macro', 'run-macro',
+                                     'set-cmd-text']:
+                record_macro = False
```
Comment (The-Compiler) [inline qutebrowser/commands/runners.py]: I'm not 100% confident this all does the right thing with multiple mode-switching commands, but I *think* it's fine.

File: qutebrowser/keyinput/macros.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Keyboard macro system."""
+
+from qutebrowser.commands import cmdexc, cmdutils, runners
+from qutebrowser.utils import message, objreg, usertypes
+
+
+class MacroRecorder:
+    """An object for recording and running keyboard macros."""
```
Comment (The-Compiler) [inline qutebrowser/keyinput/macros.py]: nitpick: Please add a blank line before the docstring.

Also, please add an `Attributes:` section describing the attributes - it took me some minutes to understand why `macro_count` is needed.

File: qutebrowser/keyinput/macros.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+    def __init__(self):
+        self.macro = {}
```
Comment (The-Compiler) [inline qutebrowser/keyinput/macros.py]: This should probably be named `macros`

File: qutebrowser/keyinput/macros.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        self.recording_macro = None
+        self.macro_count = {}
```
Comment (The-Compiler) [inline qutebrowser/keyinput/macros.py]: Those attributes are all not needed from the outside, correct? Then they should be named `self._macro_count` etc. (to make them "private")

File: qutebrowser/keyinput/macros.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    def record_macro_command(self, win_id, register=None):
+        """Start or stop recording a macro.
+
+        Args:
+            register: Which register to store the macro in.
+        """
+        if self.recording_macro is None:
+            if register is None:
+                mode_manager = objreg.get('mode-manager', scope='window',
+                                          window=win_id)
```
Comment (The-Compiler) [inline qutebrowser/keyinput/macros.py]: You can use `modeman.instance(win_id)` from the `keyinput.modeman` module instead.

File: qutebrowser/keyinput/macros.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                mode_manager.enter(usertypes.KeyMode.record_macro,
+                                   'record_macro')
+            else:
+                self.record_macro(register)
+        else:
+            message.info("Macro recorded.")
```
Comment (The-Compiler) [inline qutebrowser/keyinput/macros.py]: Maybe add the register name here?

File: qutebrowser/keyinput/macros.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            self.recording_macro = None
+
+    def record_macro(self, register):
+        """Start recording a macro."""
+        message.info("Recording macro...")
```
Comment (The-Compiler) [inline qutebrowser/keyinput/macros.py]: Maybe add the register name here?

File: qutebrowser/keyinput/macros.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        """Run a recorded macro.
+
+        Args:
+            count: How many times to run the macro.
+            register: Which macro to run.
+        """
+        self.macro_count[win_id] = count
+        if register is None:
+            mode_manager = objreg.get('mode-manager', scope='window',
+                                      window=win_id)
```
Comment (The-Compiler) [inline qutebrowser/keyinput/macros.py]: `modeman.instance` too

File: qutebrowser/keyinput/modeparsers.py [hunk 9]
```diff
@@ -294,23 +298,32 @@ def handle(self, e):
...
+                tabbed_browser.set_mark(key)
+            elif self._mode == usertypes.KeyMode.jump_mark:
+                tabbed_browser.jump_mark(key)
+            elif self._mode == usertypes.KeyMode.record_macro:
+                macro_recorder.record_macro(key)
+            elif self._mode == usertypes.KeyMode.run_macro:
+                macro_recorder.run_macro(self._win_id, key)
+            else:
+                raise ValueError(
+                    "{} is not a valid register key".format(self._mode))
```
Comment (The-Compiler) [inline qutebrowser/keyinput/modeparsers.py]: register key -> register mode?

File: qutebrowser/commands/runners.py [hunk 10]
```diff
@@ -259,19 +259,33 @@ def run(self, text, count=None):
...
         """
-        for result in self.parse_all(text):
-            mode_manager = objreg.get('mode-manager', scope='window',
-                                      window=self._win_id)
-            cur_mode = mode_manager.mode
+        record_last_command = True
+        record_macro = True
+
+        mode_manager = objreg.get('mode-manager', scope='window',
+                                  window=self._win_id)
```
Comment (The-Compiler) [inline qutebrowser/commands/runners.py]: Didn't see this before, but you can use `modeman.instance` here too.

File: qutebrowser/keyinput/macros.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    """
+
+    def __init__(self):
+        self._macros = {}
+        self._recording_macro = None
+        self._macro_count = {}
+
+    @cmdutils.register(instance='macro-recorder', name='record-macro')
+    @cmdutils.argument('win_id', win_id=True)
+    @cmdutils.argument('register')
```
Comment (The-Compiler) [inline qutebrowser/keyinput/macros.py]: No need for `@cmdutils.argument` when you don't need to provide more information - it'll be picked up automatically.

File: qutebrowser/keyinput/macros.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    def record_macro(self, register):
+        """Start recording a macro."""
+        message.info("Recording macro '{}'...".format(register))
+        self._macros[register] = []
+        self._recording_macro = register
+
+    @cmdutils.register(instance='macro-recorder', name='run-macro')
+    @cmdutils.argument('win_id', win_id=True)
+    @cmdutils.argument('count', count=True)
+    @cmdutils.argument('register')
```
Comment (The-Compiler) [inline qutebrowser/keyinput/macros.py]: Same as above

File: qutebrowser/keyinput/macros.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        """Run a recorded macro."""
+        if register not in self._macros:
+            raise cmdexc.CommandError(
+                "No macro recorded in '{}'!".format(register))
+        commandrunner = runners.CommandRunner(win_id)
+        for _ in range(self._macro_count[win_id]):
+            for cmd in self._macros[register]:
+                commandrunner.run_safely(*cmd)
+
+    def record(self, text, count):
```
Comment (The-Compiler) [inline qutebrowser/keyinput/macros.py]: Maybe call this `record_command` instead? It's a bit ambiguous now.

**Comment by The-Compiler** [comment]
Thanks!

I moved the tests to `keyinput.feature` since I think they fit better there and `misc.feature` has too much stuff already anyways :laughing:

After merging I also noticed an exception when e.g. using hints via a macro - fixed in 2ef85d6c3561bcc856274a5c823192e1f40216df.

**Pull request merged**